include/nbl/builtin/hlsl/bxdf/fresnel.hlsl

```diff
@@ -646,11 +640,80 @@ struct Iridescent
         return hlsl::max(colorspace::scRGB::FromXYZ(I), hlsl::promote<vector_type>(0.0)) * hlsl::promote<vector_type>(0.5);
     }
 
-    scalar_type getRefractionOrientedEta() NBL_CONST_MEMBER_FUNC { return eta23[0]; }
-    OrientedEtaRcps<monochrome_type> getOrientedEtaRcps() NBL_CONST_MEMBER_FUNC
+    scalar_type Dinc;       // thickness of thin film in nanometers, rec. 100-25000nm
+    vector_type thinFilmIor;
+    vector_type eta12;      // outside (usually air 1.0) -> thin-film IOR
+    vector_type eta23;      // thin-film -> base material IOR
+    vector_type etak23;     // thin-film -> complex component, k==0 makes dielectric
+};
+}
+
+template<typename T>
+NBL_PARTIAL_REQ_TOP(concepts::FloatingPointLikeVectorial<T>)
+struct Iridescent<T, false NBL_PARTIAL_REQ_BOT(concepts::FloatingPointLikeVectorial<T>) >
+{
+    using this_t = Iridescent<T,false>;
+    using scalar_type = typename vector_traits<T>::scalar_type;
+    using vector_type = T;  // assert dim==3?
+    using eta_type = vector_type;
+
+    static this_t create(scalar_type Dinc, vector_type ior1, vector_type ior2, vector_type ior3, vector_type iork3)
+    {
+        this_t retval;
+        retval.helper.Dinc = Dinc;
+        retval.helper.thinFilmIor = ior2;
+        retval.helper.eta12 = ior2/ior1;
+        retval.helper.eta23 = ior3/ior2;
+        retval.helper.etak23 = iork3/ior2;
+        return retval;
+    }
+
+    T operator()(const scalar_type clampedCosTheta)
+    {
+        return helper.__call(clampedCosTheta);
+    }
+
+    OrientedEtaRcps<eta_type> getOrientedEtaRcps() NBL_CONST_MEMBER_FUNC
+    {
+        OrientedEtaRcps<eta_type> rcpEta;
+        rcpEta.value = hlsl::promote<eta_type>(1.0) / helper.eta23;
+        rcpEta.value2 = rcpEta.value * rcpEta.value;
+        return rcpEta;
+    }
+
+    impl::iridescent_helper<T,false> helper;
+};
+
+template<typename T>
+NBL_PARTIAL_REQ_TOP(concepts::FloatingPointLikeVectorial<T>)
+struct Iridescent<T, true NBL_PARTIAL_REQ_BOT(concepts::FloatingPointLikeVectorial<T>) >
+{
+    using this_t = Iridescent<T,true>;
+    using scalar_type = typename vector_traits<T>::scalar_type;
+    using vector_type = T;  // assert dim==3?
+    using eta_type = vector<scalar_type, 1>;
+
+    static this_t create(scalar_type Dinc, vector_type ior1, vector_type ior2, vector_type ior3)
     {
-        OrientedEtaRcps<monochrome_type> rcpEta;
-        rcpEta.value = hlsl::promote<monochrome_type>(1.0) / eta23[0];
+        this_t retval;
+        retval.helper.Dinc = Dinc;
+        retval.helper.thinFilmIor = ior2;
+        retval.helper.eta12 = ior2/ior1;
+        retval.helper.eta23 = ior3/ior2;
+        retval.helper.etak23 = hlsl::promote<vector_type>(0.0);
```

Comment:
I think you could get this https://github.com/Devsh-Graphics-Programming/Nabla/pull/918#discussion_r2486512597 done via CRTP where you feed this class into the `helper` class and it has `getEta12()` etc getters

or the other way round, a base class with getters and it goes into helper class without Curious Recurrence